File: Finales/ejercicios_sueltos_finales.py
#EJERCICIO 1:
def nuevaMatriz (sumar: list[int], m:list[list[int]]) -> list[list[int]]:
    resultado: list[list[int]] = []
    fila: list[int] = []
    i:int=0
    e:int=0

    for f in m:
        while i<len(sumar):
              e+= f[i] + sumar[i]
              fila.append(e)
              i+=1
              e=0
        resultado.append(fila)
        fila=[]
        i=0
        e=0

    return resultado          


m = [[1,4,3,2],
     [6,1,8,9],
     [2,9,7,4],
     [3,2,3,3]
     ]

s = [3,2,8,4]

print(nuevaMatriz(s,m))





#EJERCICIO 2:
def esDamero (m:list[list[int]]) -> bool:
    i:int=0
    c:int=0
    columna:list[int]=[]
    for fila in m:
         if len(fila)==8:
            while (i+1) < len(fila):
                  if fila[i]==1:
                     if fila[i+1]==0:
                        res=True
                     else:
                        res=False
                        break
                  elif fila[i]==0 :
                     if fila[i+1]==1:
                        res=True
                     else:
                        res=False
                        break 
                  i+=1
         else:
             res=False
             break
         i=0
    while c<8:  
        for fila in m:
            columna.append(fila[c])
        if len(columna)==8:
           while (i+1) < len(columna):
                  if columna[i]==1:
                     if columna[i+1]==0:
                        res=True
                     else:
                        res=False
                        break
                  elif columna[i]==0 :
                     if fila[i+1]==1:
                        res=True
                     else:
                        res=False
                        break 
                  i+=1
        else:
             res=False 
             break    
        columna=[]
        c+=1
        i=0

    return res    
                                      
                        
tablero = [[1,0,1,0,1,0,1,0],
     [0,1,0,1,0,1,0,1],
     [1,0,1,0,1,0,1,0],
     [0,1,0,1,0,1,0,1],
     [1,0,1,0,1,0,1,0],
     [0,1,0,1,0,1,0,1],
     [1,0,1,0,1,0,1,0],
     [0,1,0,1,0,1,0,1]
]

falso_tablero = [[0,1,0],
                 [1,0,1],
                 [0,1,0]
                 ]

print(esDamero(tablero))
print(esDamero(falso_tablero))





#EJERCICIO 3:
def quien_gano_el_tateti_facilito (tablero: list[list[chr]]) -> bool:
    c:int=0
    columna: list[chr]=[]
    hay_tres_x:list[bool]=[]
    hay_tres_o:list[bool]=[]
    i:int=0

    while c<len(tablero):
          for fila in tablero:
              columna.append(fila[c])
          while (i+2)<len(tablero):
              if columna[i]=='X':
                 if (columna[i+1]=='X') and (columna[i+2]=='X'):
                     res_x =True
                     hay_tres_x.append(res_x)
              elif columna[i]=='O':
                 if (columna[i+1]=='O') and (columna[i+2]=='O'):
                     res_o =True
                     hay_tres_o.append(res_o)  
              i+=1
              res_x = False
              res_o = False
          columna = []
          c+=1
          i=0

    if (True in hay_tres_x) and (True in hay_tres_o):
          res = 3
    elif (True in hay_tres_x) and not (True in hay_tres_o):
          res = 1
    elif not (True in hay_tres_x) and (True in hay_tres_o):
          res = 2        
    elif not (True in hay_tres_x) and not (True in hay_tres_o):
          res = 0

    return res         

t = [['X',' ',' ',' ',' '],
     ['X',' ','O',' ',' '],
     [' ',' ','O',' ',' '],
     [' ',' ','O',' ',' '],
     [' ',' ',' ',' ',' ']
     ]

print(quien_gano_el_tateti_facilito(t))
                       
                

File: Finales/ejercicios_sueltos_finales.hs
--EJERCICIO 1:
nuevaMatriz :: [Int] -> [[Int]] -> [[Int]]
nuevaMatriz (x:xs) [y] = [sumaElementos (x:xs) y]
nuevaMatriz (x:xs) (y:ys) = [sumaElementos (x:xs) y] ++ nuevaMatriz (x:xs) (ys)
nuevaMatriz (x:xs) [] = []

sumaElementos :: [Int] -> [Int] -> [Int]
sumaElementos [x] [y] = [(x+y)]
sumaElementos (x:xs) (y:ys) = [(x+y)] ++ sumaElementos (xs) (ys)

--m = [[1,4,3,2],
--     [6,1,8,9],
--     [2,9,7,4],
--     [3,2,3,3]
--     ]

--s = [3,2,8,4]





--EJERCICIO 2:
esDamero :: [[Int]] -> Bool
esDamero (x:xs) | filasValidas (x:xs) && columnasValidas (x:xs) && (longitud (x:xs)==8) =True
                | otherwise =False

longitud :: [[Int]] -> Int
longitud [x] = 1
longitud (x:xs) = 1 + longitud (xs)  

secuenciaValida :: [Int] -> Bool
secuenciaValida (x:y:[]) | x==1 && y==0 =True
                         | x==0 && y==1 =True
                         | otherwise =False
secuenciaValida (x:y:xs) | x==1 && y==0 =secuenciaValida (y:xs)
                         | x==0 && y==1 =secuenciaValida (y:xs)
                         | otherwise =False

filasValidas ::[[Int]] -> Bool
filasValidas [x] | secuenciaValida x =True
                 | otherwise =False
filasValidas (x:xs) | secuenciaValida x =filasValidas (xs)
                    | otherwise =False

quitarPrimerElemento :: [Int] -> [Int]
quitarPrimerElemento (x:xs) = xs
quitarPrimerElemento [] = []

quitarPrimeraColumna :: [[Int]] -> [[Int]]
quitarPrimeraColumna [x] = [quitarPrimerElemento x]
quitarPrimeraColumna (x:xs) = [quitarPrimerElemento x] ++ quitarPrimeraColumna (xs)
quitarPrimeraColumna [] = []                    

primeraColumna :: [[Int]] -> [Int]
primeraColumna [] = []
primeraColumna [x] = [head x]
primeraColumna (x:xs) = [head x] ++ primeraColumna (xs)

columnas :: [[Int]] -> [[Int]]
columnas [] = []
columnas ([]:_) = []
columnas (x:xs) = [primeraColumna (x:xs)] ++ columnas (quitarPrimeraColumna(x:xs))

columnasValidas :: [[Int]] -> Bool
columnasValidas (x:xs) | filasValidas (columnas (x:xs)) =True
                       | otherwise =False


--tablero = [[1,0,1,0,1,0,1,0],
--     [0,1,0,1,0,1,0,1],
--     [1,0,1,0,1,0,1,0],
--     [0,1,0,1,0,1,0,1],
--     [1,0,1,0,1,0,1,0],
--     [0,1,0,1,0,1,0,1],
--     [1,0,1,0,1,0,1,0],
--     [0,1,0,1,0,1,0,1]
--]

--falso_tablero = [[0,1,0],
--                 [1,0,1],
--                 [0,1,0]
--                 ]